File: app/src/main/java/uk/co/ourfriendirony/medianotifier/clients/AbstractClient.java
package uk.co.ourfriendirony.medianotifier.clients;

import android.util.Log;

import java.io.IOException;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public abstract class AbstractClient {
    private static final int SLEEP = 1;

    protected String httpGetRequest(String url) throws IOException {
        OkHttpClient client = new OkHttpClient();
        Request req = new Request.Builder()
                .url(url)
                .addHeader("User-Agent", "MediaNotifier/1.0.0 ( ourfriendirony@gmail.com )")
                .build();
        while (true) {
            try (Response res = client.newCall(req).execute()) {
                String payload = res.body().string();
                int statusCode = res.code();
                String headers = parseHeaders(res);

                logResponse(url, payload, headers, statusCode);
                if (statusCode == 200) {
                    return payload;
                } else {
                    // TODO: This is 100% unacceptable implementation for a production release product
                    sleep(SLEEP);
                }
            }
        }
    }

    private void logResponse(String url, String payload, String headers, int statusCode) {
        Log.d("[REQUEST URL]", url);
        Log.d("[RESPONSE PAYLOAD]", payload);
        Log.d("[RESPONSE HEADERS]", headers);
        Log.d("[RESPONSE STATUSC]", String.valueOf(statusCode));
    }

    private void sleep(int time) {
        try {
            Thread.sleep(time * 1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private String parseHeaders(Response res) {
        StringBuilder builder = new StringBuilder();
        for (String name : res.headers().names()) {
            builder.append(name).append(":").append(res.headers().get(name)).append("|");
        }
        return builder.toString();
    }
}

File: app/src/main/java/uk/co/ourfriendirony/medianotifier/clients/AbstractClient.kt
package uk.co.ourfriendirony.medianotifier.clients

import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.IOException

abstract class AbstractClient {
    @Throws(IOException::class)
    protected fun httpGetRequest(url: String): String {
        val client = OkHttpClient()
        val req: Request = Request.Builder()
                .url(url)
                .addHeader("User-Agent", "MediaNotifier/1.0.0 ( ourfriendirony@gmail.com )")
                .build()
        while (true) {
            client.newCall(req).execute().use { res ->
                val payload = res.body!!.string()
                val statusCode = res.code
                val headers = parseHeaders(res)
                logResponse(url, payload, headers, statusCode)
                if (statusCode == 200) {
                    return payload
                } else {
                    // TODO: This is 100% unacceptable implementation for a production release product
                    sleep(SLEEP)
                }
            }
        }
    }

    private fun logResponse(url: String, payload: String, headers: String, statusCode: Int) {
        Log.d("[REQUEST URL]", url)
        Log.d("[RESPONSE PAYLOAD]", payload)
        Log.d("[RESPONSE HEADERS]", headers)
        Log.d("[RESPONSE STATUSC]", statusCode.toString())
    }

    private fun sleep(time: Int) {
        try {
            Thread.sleep((time * 1000).toLong())
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    private fun parseHeaders(res: Response): String {
        val builder = StringBuilder()
        for (name in res.headers.names()) {
            builder.append(name).append(":").append(res.headers[name]).append("|")
        }
        return builder.toString()
    }

    companion object {
        private const val SLEEP = 1
    }
}